<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>IMPACT-Pipeline: MSK-IMPACT assay data analysis framework to process data on LSF/SGE &mdash; IMPACT-Pipeline 1.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="IMPACT-Pipeline 1.0.1 documentation" href="#" />
    <link rel="next" title="Installation" href="install.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="install.html" title="Installation"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">IMPACT-Pipeline 1.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="impact-pipeline-msk-impact-assay-data-analysis-framework-to-process-data-on-lsf-sge">
<h1>IMPACT-Pipeline: MSK-IMPACT assay data analysis framework to process data on LSF/SGE<a class="headerlink" href="#impact-pipeline-msk-impact-assay-data-analysis-framework-to-process-data-on-lsf-sge" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body"><a class="reference external" href="https://github.com/rhshah">Ronak H Shah</a>, Donavan Cheng, <a class="reference external" href="https://github.com/ahmetz">Ahmet Zehir</a>, Aijazuddin Syed, <a class="reference external" href="https://github.com/rghu">Raghu Chandramohan</a>, Meera Prasad et.al</td>
</tr>
<tr class="field-even field"><th class="field-name">Contact:</th><td class="field-body"><a class="reference external" href="mailto:rons&#46;shah&#37;&#52;&#48;gmail&#46;com">rons<span>&#46;</span>shah<span>&#64;</span>gmail<span>&#46;</span>com</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">Source code:</th><td class="field-body"><a class="reference external" href="http://github.com/rhshah/IMPACT-Pipeline">http://github.com/rhshah/IMPACT-Pipeline</a></td>
</tr>
<tr class="field-even field"><th class="field-name">License:</th><td class="field-body"><a class="reference external" href="http://www.apache.org/licenses/LICENSE-2.0">Apache License 2.0</a></td>
</tr>
</tbody>
</table>
<p>IMPACT-Pipeline is a Perl/Python/R command-line software toolkit to process data, call somatic aberrations (SNV,INDELS,SV,CNV) generated by NGS based MSK-IMPACT assay.
It is designed for use with hybrid capture, including both whole-exome and custom target panels, and short-read sequencing platforms such as Illumina.</p>
<a class="reference external image-reference" href="http://dx.doi.org/10.5281/zenodo.32716"><img src="https://zenodo.org/badge/doi/10.5281/zenodo.32716.svg" /></a>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="requirements.html">Requirements</a><ul>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#required-perl-packages">Required Perl Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#required-python-packages">Required Python Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#required-r-packages">Required R Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#purpose-and-tools-used">Purpose and Tools Used</a></li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#inside-the-config-file">Inside the config file</a><ul>
<li class="toctree-l3"><a class="reference internal" href="requirements.html#locations">Locations</a></li>
<li class="toctree-l3"><a class="reference internal" href="requirements.html#parameters">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="requirements.html#versions">Versions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#description-for-title-file-txt">Description for title_file.txt</a></li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#description-for-samplesheet-csv">Description for SampleSheet.csv</a></li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#description-for-adaptor-file-in-the-configuration-file">Description for adaptor file in the configuration file</a></li>
<li class="toctree-l2"><a class="reference internal" href="requirements.html#description-for-barcode-file-in-the-configuration-file">Description for barcode file in the configuration file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="usage.html#quick-usage">Quick Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html#detailed-usage">Detailed Usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="usage.html#what-does-each-number-represent">What does each number represent</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage.html#using-different-process-to-run-pipeline">Using different Process to run Pipeline</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage.html#shell-script-to-run-pipeline">Shell Script to run pipeline</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="subscripts.html">Description of sub-scripts</a><ul>
<li class="toctree-l2"><a class="reference internal" href="subscripts.html#script-in-the-bin-folder">Script in the bin folder</a><ul>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#compiling-qc-metrics-generating-qc-report-and-running-copynumber-analysis">Compiling QC metrics, Generating QC-Report and running copynumber analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#genotyping-variants-across-multiple-samples">Genotyping Variants across multiple samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#running-indel-realignment-using-abra">Running Indel Realignment using ABRA</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#call-indels-25bp-using-pindel">Call Indels &gt; 25bp using Pindel</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="subscripts.html#script-in-the-support-scripts-folder">Script in the support-scripts folder</a><ul>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#calculate-intervals-from-bam-file-that-have-some-minimum-coverage">Calculate intervals from bam file that have some minimum coverage</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#annotating-variants-in-merged-variant-file">Annotating variants in merged variant file</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#filter-variants-after-annotation">Filter variants after annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#filter-indels-from-somaticindeldetector-before-genotyping">Filter indels from SomaticIndelDetector before genotyping</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#filter-snv-from-mutect-before-genotyping">Filter snv from MuTect before genotyping</a></li>
<li class="toctree-l3"><a class="reference internal" href="subscripts.html#filter-indels-from-pindel-before-genotyping">Filter indels from PINDEL before genotyping</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="filedescriptions.html">File Description</a><ul>
<li class="toctree-l2"><a class="reference internal" href="filedescriptions.html#qc-metrics-files">QC Metrics Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-metrics-pdf">Proj_*_All_Metrics.pdf</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-basequalities-txt">Proj_*_ALL_basequalities.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-canonical-exoncoverage-txt">Proj_*_ALL_Canonical_exoncoverage.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-exoncoverage-txt">Proj_*_ALL_exoncoverage.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-exonnomapqcoverage-txt">Proj_*_ALL_exonnomapqcoverage.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-fpavghom-txt">Proj_*_ALL_FPavgHom.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-fpcresultsunmatch-txt">Proj_*_ALL_FPCResultsUnMatch.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-fpcresultsunmismatch-txt">Proj_*_ALL_FPCResultsUnMismatch.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-fpcsummary-txt">Proj_*_ALL_FPCsummary.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-fphet-txt">Proj_*_ALL_FPhet.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-fpsummary-txt">Proj_*_ALL_FPsummary.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-gcbias-txt">Proj_*_ALL_gcbias.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-genecoverage-txt">Proj_*_ALL_genecoverage.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-genotypehotspotnormals-txt">Proj_*_ALL_genotypehotspotnormals.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-hsmetrics-txt">Proj_*_ALL_HSmetrics.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-insertsizemetrics-txt">Proj_*_ALL_insertsizemetrics.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-intervalnomapqcoverage-loess-txt">Proj_*_ALL_intervalnomapqcoverage_loess.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-intervalnomapqcoverage-txt">Proj_*_ALL_intervalnomapqcoverage.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-orgbasequalities-txt">Proj_*_ALL_orgbasequalities.txt</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="filedescriptions.html#copy-number-files">Copy Number Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-all-copynumber-seg">Proj_*_ALL_copynumber.seg</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-copynumber-segclusp-genes-txt">Proj_*_copynumber_segclusp.genes.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-copynumber-segclusp-intragenic-txt">Proj_*_copynumber_segclusp.intragenic.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-copynumber-segclusp-pdf">Proj_*_copynumber_segclusp.pdf</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-copynumber-segclusp-probes-txt">Proj_*_copynumber_segclusp.probes.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-discrete-cna-txt">Proj_*_discrete_CNA.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-loessnorm-pdf">Proj_*_loessnorm.pdf</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="filedescriptions.html#structural-variant-files">Structural Variant Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allannotatedsvs-txt">Proj_*_AllAnnotatedSVs.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allannotatedsvs-xlsx">Proj_*_AllAnnotatedSVs.xlsx</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="filedescriptions.html#per-sample-files">Per Sample Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#s-proj-copynumber-seg">s_*_Proj_*_copynumber.seg</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#s-proj-mrg-cl-aln-srt-md-ir-br-canonical-exon-covg-sample-interval-summary">s_*_Proj_*_mrg_cl_aln_srt_MD_IR_BR.canonical.exon.covg.sample_interval_summary</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#s-proj-mrg-cl-aln-srt-md-ir-br-gene-nomapq-covg-sample-interval-summary">s_*_Proj_*_mrg_cl_aln_srt_MD_IR_BR.gene_nomapq.covg.sample_interval_summary</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#s-proj-mrg-cl-aln-srt-md-ir-br-gene-covg-sample-interval-summary">s_*_Proj_*_mrg_cl_aln_srt_MD_IR_BR.gene.covg.sample_interval_summary</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#s-proj-mrg-cl-aln-srt-md-ir-br-target-covg">s_*_Proj_*_mrg_cl_aln_srt_MD_IR_BR.target.covg</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#s-proj-mrg-cl-aln-srt-md-ir-br-tiling-nomapq-covg-sample-interval-summary">s_*_Proj_*_mrg_cl_aln_srt_MD_IR_BR.tiling_nomapq.covg.sample_interval_summary</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#s-proj-mrg-cl-aln-srt-md-ir-br-tiling-covg-sample-interval-summary">s_*_Proj_*_mrg_cl_aln_srt_MD_IR_BR.tiling.covg.sample_interval_summary</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="filedescriptions.html#sample-info-file">Sample Info File</a><ul>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-title-file-txt">Proj_*_title_file.txt</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="filedescriptions.html#variant-files">Variant Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#annotated-exonic-variants-txt">annotated_exonic_variants.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotated-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotated.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatedexonic-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedExonic.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatedexonic-dropped-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedExonic.Dropped.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatedexonic-filtered-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedExonic.Filtered.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatedsilent-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedSilent.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatedsilent-dropped-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedSilent.Dropped.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatedsilent-filtered-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedSilent.Filtered.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatednonpanelexonic-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedNonPanelExonic.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatednonpanelexonic-dropped-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedNonPanelExonic.Dropped.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatednonpanelexonic-filtered-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedNonPanelExonic.Filtered.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatednonpanelsilent-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedNonPanelSilent.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatednonpanelsilent-dropped-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedNonPanelSilent.Dropped.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotatednonpanelsilent-filtered-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotatedNonPanelSilent.Filtered.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-annovarannotated-all-filtered-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_annovarAnnotated_All_Filtered.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="filedescriptions.html#proj-allsomaticmutindel-withalleledepth-mergeddnp-txt">Proj_*_AllSomaticMutIndel_withAlleleDepth_mergedDNP.txt</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="citation">
<h1>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h1>
<p>If you use this software in a publication, please cite our paper</p>
<p><a class="reference external" href="http://www.sciencedirect.com/science/article/pii/S1525157815000458">Memorial Sloan Kettering-Integrated Mutation Profiling of Actionable Cancer Targets (MSK-IMPACT): A Hybridization Capture-Based Next-Generation Sequencing Clinical Assay for Solid Tumor Molecular Oncology</a>.</p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">IMPACT-Pipeline: MSK-IMPACT assay data analysis framework to process data on LSF/SGE</a></li>
<li><a class="reference internal" href="#citation">Citation</a></li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href="install.html"
                        title="next chapter">Installation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="install.html" title="Installation"
             >next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">IMPACT-Pipeline 1.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Ronak H Shah, Donavan Cheng, Ahmet Zehir, Aijazuddin Syed, Raghu Chandramohan, Meera Prasad.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>